File: ipaacalib/python/src/notifier.py
'''
Created on Dec 20, 2012

@author: hvanwelbergen
'''
from __future__ import with_statement
import threading

from ipaaca import IUEventType
from ipaaca import Message
class ComponentNotifier(object):
    '''
    classdocs
    '''

    NOTIFY_CATEGORY = "componentNotify";
    SEND_CATEGORIES = "send_categories";
    RECEIVE_CATEGORIES = "recv_categories";
    STATE = "state";
    NAME = "name";
    FUNCTION = "function";    
    
    def __init__(self, componentName, componentFunction, sendCategories, receiveCategories, outBuffer, inBuffer):
        '''
        Constructor
        '''
        self.componentName = componentName
        self.componentFunction = componentFunction
        self.sendCategories = frozenset(sendCategories)
        self.receiveCategories = frozenset(receiveCategories)
        self.inBuffer = inBuffer
        self.outBuffer = outBuffer
        self.initialized = False
        self.notificationHandlers = []
        self.initializeLock = threading.Lock()
        self.notificationHandlerLock = threading.Lock()
        self.submitLock = threading.Lock()        
    
    def _submit_notify(self, isNew):
        with self.submitLock:
            notifyIU = Message(ComponentNotifier.NOTIFY_CATEGORY)
            notifyIU.payload[ComponentNotifier.NAME] = self.componentName
            notifyIU.payload[ComponentNotifier.FUNCTION] = self.componentFunction
            notifyIU.payload[ComponentNotifier.SEND_CATEGORIES] = ",".join(self.sendCategories)
            notifyIU.payload[ComponentNotifier.RECEIVE_CATEGORIES] =  ",".join(self.receiveCategories)
            notifyIU.payload[ComponentNotifier.STATE] = "new" if isNew else "old"                
            
            self.outBuffer.add(notifyIU)
    
    def _handle_iu_event(self, iu, event_type, local):
        print("handle, iuname:"+iu.payload[ComponentNotifier.NAME]+" component name: "+self.componentName+" state "+iu.payload[ComponentNotifier.STATE])
        if iu.payload[ComponentNotifier.NAME] == self.componentName:
            return
        with self.notificationHandlerLock:
            for h in self.notificationHandlers:
                h(iu, event_type, local)
        if iu.payload[ComponentNotifier.STATE] == "new":
            print("submitting")
            self._submit_notify(False)            

    def add_notification_handler(self, handler):
        with self.notificationHandlerLock:
            self.notificationHandlers.append(handler)
                
    def initialize(self):
        with self.initializeLock:
            if (not self.initialized):
                self.inBuffer.register_handler(self._handle_iu_event, [IUEventType.MESSAGE], ComponentNotifier.NOTIFY_CATEGORY)
                self._submit_notify(True)
                self.initialized = True
            
    
    
            
        

File: ipaacalib/cpp/src/util/notifier.cc
#include <ipaaca/util/notifier.h>

namespace ipaaca {
namespace util {

ComponentNotifier::~ComponentNotifier() 
{
	if (initialized) {
		submit_notify(_IPAACA_COMP_NOTIF_STATE_DOWN);
	}
}

ComponentNotifier::ComponentNotifier(const std::string& componentName, const std::string& componentFunction, const std::set<std::string>& sendCategories, const std::set<std::string>& recvCategories)
: initialized(false), name(componentName), function(componentFunction)
{
	send_categories = ipaaca::str_join(sendCategories, ",");
	recv_categories = ipaaca::str_join(recvCategories, ",");
	// create private in/out buffer pair since none was specified
	out_buf = ipaaca::OutputBuffer::create(componentName);
	in_buf = ipaaca::InputBuffer::create(componentName, _IPAACA_COMP_NOTIF_CATEGORY);
}

ComponentNotifier::ComponentNotifier(const std::string& componentName, const std::string& componentFunction, const std::set<std::string>& sendCategories, const std::set<std::string>& receiveCategories, ipaaca::OutputBuffer::ptr outBuf, ipaaca::InputBuffer::ptr inBuf)
: initialized(false), name(componentName), function(componentFunction), out_buf(outBuf), in_buf(inBuf)
{
	send_categories = ipaaca::str_join(sendCategories, ",");
	recv_categories = ipaaca::str_join(receiveCategories, ",");
}

ComponentNotifier::ptr ComponentNotifier::create(const std::string& componentName, const std::string& componentFunction, const std::set<std::string>& sendCategories, const std::set<std::string>& recvCategories)
{
	return ComponentNotifier::ptr(new ComponentNotifier(componentName, componentFunction, sendCategories, recvCategories));
}
ComponentNotifier::ptr ComponentNotifier::create(const std::string& componentName, const std::string& componentFunction, const std::set<std::string>& sendCategories, const std::set<std::string>& recvCategories, ipaaca::OutputBuffer::ptr outBuf, ipaaca::InputBuffer::ptr inBuf)
{
	return ComponentNotifier::ptr(new ComponentNotifier(componentName, componentFunction, sendCategories, recvCategories, outBuf, inBuf));
}

void ComponentNotifier::handle_iu_event(IUInterface::ptr iu, IUEventType event_type, bool local)
{
	//std::cout << "handle_iu_event: got an event" << std::endl;
	if ((event_type == IU_ADDED) || (event_type == IU_UPDATED) || (event_type == IU_MESSAGE)) {
		Locker locker(lock);
		std::string cName = iu->payload()[_IPAACA_COMP_NOTIF_NAME];
		std::string cState = iu->payload()[_IPAACA_COMP_NOTIF_STATE];
		if (cName != name) {
			//std::cout << " handle_iu_event: calling notification handlers" << std::endl;
			// call all registered notification handlers
			for (std::vector<IUEventHandlerFunction>::iterator it = _handlers.begin(); it != _handlers.end(); ++it) {
				(*it)(iu, event_type, local);
			}
			// send own info only if the remote component is a newly initialized one
			if (cState=="new") {
				//std::cout << " handle_iu_event: Submitting own notification to new remote end" << std::endl;
				submit_notify(_IPAACA_COMP_NOTIF_STATE_OLD);
			}
		}
	}
}

void ComponentNotifier::add_notification_handler(ipaaca::IUEventHandlerFunction function)
{
	Locker locker(lock);
	_handlers.push_back(function);
}

void ComponentNotifier::submit_notify(const std::string& current_state)
{
	ipaaca::Message::ptr iu = ipaaca::Message::create(_IPAACA_COMP_NOTIF_CATEGORY);
	iu->payload()[_IPAACA_COMP_NOTIF_NAME] = name;
	iu->payload()[_IPAACA_COMP_NOTIF_STATE] = current_state;
	iu->payload()[_IPAACA_COMP_NOTIF_FUNCTION] = function;
	iu->payload()[_IPAACA_COMP_NOTIF_SEND_CATS] = send_categories;
	iu->payload()[_IPAACA_COMP_NOTIF_RECV_CATS] = recv_categories;
	out_buf->add(iu);
	//LOG_IPAACA_CONSOLE( "Sending a ComponentNotify: " << name << " " << function << " " << current_state << " " << send_categories << " " << recv_categories )
}

void ComponentNotifier::initialize() {
	Locker locker(lock);
	if (!initialized) {
		initialized = true;
		in_buf->register_handler(boost::bind(&ComponentNotifier::handle_iu_event, this, _1, _2, _3));
		submit_notify(_IPAACA_COMP_NOTIF_STATE_NEW);
	}
}

}}

